Move content from coordinate reference to guide (#6881)
sunpy/sunpy: docs/guide/coordinates/index.rst

.. _sunpy-coordinates:

***********
Coordinates
***********

The `sunpy.coordinates` sub-package contains:

* A robust framework for working with solar-physics coordinate systems
* Functions to obtain the locations of solar-system bodies (`sunpy.coordinates.ephemeris`)
* Functions to calculate Sun-specific coordinate information (`sunpy.coordinates.sun`)

The SunPy coordinate framework extends the
:ref:`Astropy coordinates framework <astropy:astropy-coordinates>`.

> This page contains a short overview of how coordinates work in sunpy.

Reviewer: ```suggestion
This page contains an overview of how coordinates work in sunpy.
```

> For more detailed information see the following pages:

Reviewer: ```suggestion
See the following pages for detailed discussion of specific topics:
```

     - HGRTN
     - similar to `~sunpy.coordinates.frames.Heliocentric`
     - The axes are permuted, with HCC X, Y, Z equivalent respectively to HGRTN Y, Z, X
   * - Heliographic Carrington
     - HGC
     - `~sunpy.coordinates.frames.HeliographicCarrington`
     -
   * - Heliographic Stonyhurst
     - HGS
     - `~sunpy.coordinates.frames.HeliographicStonyhurst`
     -
   * - Helioprojective Cartesian
     - HPC
     - `~sunpy.coordinates.frames.Helioprojective`
     -
   * - Geocentric Earth Equatorial (Mean)
     - GEI
     - `~sunpy.coordinates.frames.GeocentricEarthEquatorial`
     -
   * - Geographic
     - GEO
     - Astropy's `~astropy.coordinates.ITRS`
     - The precise geographic definitions may differ
   * - Geocentric Solar Ecliptic
     - GSE
     - `~sunpy.coordinates.frames.GeocentricSolarEcliptic`
     -

For a description of these coordinate systems,
see `Thompson (2006) <https://doi.org/10.1051/0004-6361:20054262>`_
and `Franz & Harper (2002) <https://doi.org/10.1016/S0032-0633(01)00119-2>`_
(and `corrected version <https://www2.mps.mpg.de/homes/fraenz/systems/systems3art/systems3art.html>`_).

Getting Started
===============

> 

Author: Is this something you'd like me to do in this PR?

> 

Reviewer 2: No, I don't think so. See my comment below for reasoning.

The `~sunpy.coordinates.frames.Helioprojective`, `~sunpy.coordinates.frames.Heliocentric`
and `~sunpy.coordinates.frames.HeliographicCarrington` frames are defined by the location of
the observer. For example in `~sunpy.coordinates.frames.Helioprojective` the
observer is at the origin of the coordinate system. This information is encoded
in such observer-based frames as the ``observer`` attribute, which is itself an instance of the
`~sunpy.coordinates.frames.HeliographicStonyhurst` frame.
The ``observer`` can be a string for a solar-system body (e.g., "earth" or
"mars"), and
`~sunpy.coordinates.ephemeris.get_body_heliographic_stonyhurst` will be used
with the specified ``obstime`` to fully specify the observer location.  If
the observer location is not fully specified, or not present at all, most
transformations cannot be performed.
The location of the observer is automatically populated from meta data when
coordinate frames are created using map.

In the case of `~sunpy.coordinates.frames.HeliographicCarrington`, one can specify ``observer='self'`` to indicate that the coordinate itself should be used as the observer for defining the coordinate frame.

> It is possible to convert from a `~sunpy.coordinates.frames.Helioprojective`
> frame with one observer location to another
> `~sunpy.coordinates.frames.Helioprojective` frame with a different observer
> location, by converting through `~sunpy.coordinates.frames.HeliographicStonyhurst`, this
> does involve making an assumption of the radius of the Sun to calculate the
> position on the solar sphere. The conversion can be performed as follows::

Reviewer: ```suggestion
It is possible to convert from a `~sunpy.coordinates.frames.Helioprojective` frame with one observer location to another `~sunpy.coordinates.frames.Helioprojective` frame with a different observer location.
The transformation requires the coordinate to be 3D, so if the initial coordinate is only 2D, the default assumption maps that 2D coordinate to the surface of the Sun, as defined by the ``rsun`` frame attribute.
The conversion can be performed as follows::
```

  # Input coordinate
  >>> hpc1 = SkyCoord(0*u.arcsec, 0*u.arcsec, observer="earth", obstime="2017-07-26", frame=frames.Helioprojective)

  Define a new Helioprojective frame with a different observer.
  >>> import sunpy.coordinates
  >>> hpc_out = sunpy.coordinates.Helioprojective(observer="venus", obstime="2017-07-26")

  Perform the transformation from one to the other.
  >>> hpc2 = hpc1.transform_to(hpc_out)

An example with two maps, named ``aia`` and ``stereo``::

  >>> hpc1 = SkyCoord(0*u.arcsec, 0*u.arcsec, frame=aia.coordinate_frame)  # doctest: +SKIP
  >>> hpc2 = hpc1.transform_to(stereo.coordinate_frame)  # doctest: +SKIP

Design of the Coordinates Sub-Package

> =====================================

Reviewer 2: Ok in my revised understanding of what is meant by "Guide" in the context of this PR, I think it makes sense to leave this as is.

This sub-package works by defining a collection of ``Frames``
(`sunpy.coordinates.frames`), which exists on a transformation graph, where the
transformations between the coordinate frames are then defined and registered

> with the transformation graph (`sunpy.coordinates.transformations`). It is also

Reviewer: ```suggestion
with the transformation graph (`sunpy.coordinates`). It is also
```
The transformation graph isn't shown in `sunpy.coordinates.transformations`, which will eventually be turned into a private module anyway.

system see :ref:`astropy-coordinates-overview`

Frames and SkyCoord
-------------------

The `~astropy.coordinates.SkyCoord` class is a high level wrapper around the
`astropy.coordinates` sub-package. It provides an easier way to create and transform
coordinates, by using string representations for frames rather than the classes
themselves and some other usability improvements, for more information see the
`~astropy.coordinates.SkyCoord` documentation.

The main advantage provided by `~astropy.coordinates.SkyCoord` is the support it
provides for caching Frame attributes. Frame attributes are extra data specified
with a frame, some examples in `sunpy.coordinates` are ``obstime`` or
``observer`` for observer location. Only the frames where this data is
meaningful have these attributes, i.e. only the Helioprojective frames have
``observer``. However, when you transform into another frame and then back to a
projective frame using `~astropy.coordinates.SkyCoord` it will remember the attributes previously
provided, and repopulate the final frame with them. If you were to do
transformations using the Frames alone this would not happen.

> 
> The most important implication for this in `sunpy.coordinates` is the ``rsun``
> parameter in the projective frames. If you create a projective frame with a
> ``rsun`` attribute, if you convert back to a projective frame it will be set
> correctly. It should also be noted that, if you create a Heliographic frame and
> then transform to a projective frame with an ``rsun`` attribute, it will not
> match the ``radius`` coordinate in the Heliographic frame. This is because you may
> mean to be describing a point above the defined 'surface' of the Sun.

Reviewer: ```suggestion
```
This paragraph is obsolete because `HeliographicStonyhurst` has had a `rsun` frame attribute for quite a while now.
